mm2src/coins/lightning.rs

```diff
@@ -1,100 +1,620 @@
+                    platform_coin.ticker().to_string(),
+                    "native".into(),
+                ))
+            },
+            _ => return MmError::err(LightningFromLegacyReqErr::UnexpectedMethod("".into(), "".into())),
+        };
+
+        // Channel funding transactions need to spend segwit outputs
+        // and while the witness script can be generated from pubkey and be used
+        // it's better for the coin to be enabled in segwit to check if balance is enough for funding transaction, etc...
+        if !platform_coin.as_ref().my_address.addr_format.is_segwit() {
+            return MmError::err(LightningFromLegacyReqErr::UnsupportedMode(
+                "Lightning network".into(),
+                "segwit".into(),
+            ));
+        }
+
+        let name: String = json::from_value(req["name"].clone())?;
+        if name.len() > 32 {
+            return MmError::err(LightningFromLegacyReqErr::InvalidRequest(
+                "Node name length can't be more than 32 characters".into(),
+            ));
+        }
+        let node_name = format!("{}{:width$}", name, " ", width = 32 - name.len());
+
+        let color: String = json::from_value(req["color"].clone()).unwrap_or_else(|_| "000000".into());
+        let mut node_color = [0u8; 3];
+        hex::decode_to_slice(color, &mut node_color as &mut [u8])
+            .map_to_mm(|_| LightningFromLegacyReqErr::InvalidRequest("Invalid Hex Color".into()))?;
+
+        let listening_port = json::from_value(req["port"].clone()).unwrap_or(9735);
+
+        Ok(LightningActivationParams {
+            listening_port,
+            node_name: node_name.as_bytes().try_into().expect("Node name has incorrect length"),
+            node_color,
+        })
+    }
+}
 
 #[derive(Deserialize)]
-pub struct EnableLightningRequest {
+pub struct ConnectToNodeRequest {
+    pub coin: String,
+    pub node_id: String,
+}
+
+#[cfg(target_arch = "wasm32")]
+pub async fn connect_to_lightning_node(_ctx: MmArc, _req: ConnectToNodeRequest) -> ConnectToNodeResult<String> {
+    MmError::err(ConnectToNodeError::UnsupportedMode(
+        "'connect_to_lightning_node'".into(),
+        "native".into(),
+    ))
+}
+
+/// Connect to a certain node on the lightning network.
+#[cfg(not(target_arch = "wasm32"))]
+pub async fn connect_to_lightning_node(ctx: MmArc, req: ConnectToNodeRequest) -> ConnectToNodeResult<String> {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    let ln_coin = match coin {
+        MmCoinEnum::LightningCoin(c) => c,
+        _ => return MmError::err(ConnectToNodeError::UnsupportedCoin(coin.ticker().to_string())),
+    };
+
+    let (node_pubkey, node_addr) = parse_node_info(req.node_id.clone())?;
+    let res = connect_to_node(node_pubkey, node_addr, ln_coin.peer_manager.clone()).await?;
+
+    Ok(res.to_string())
+}
+
+mod named_unit_variant {
+    named_unit_variant!(max);
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+#[serde(untagged)]
```

Comment:
Do we really require this to be untagged? We typically use adjacently tagged representation in such cases (https://serde.rs/enum-representations.html#adjacently-tagged), which seems to be handled well by GUIs.

Reply:
This allows us to use either `"amount": 3` or `"amount":"max"` in the `open_channel` request. I thought this would be more idiomatic. 

If we use adjacently tagged representation it will be `"amount": { "type":"exact", "value":3 }` or `"amount": { "type": "max" }` , but since the GUI handles this well I will fix it.

Will also remove the `named_unit_variant` macro.